### PR TITLE
Crossgen2 fixes to enable composite build with shared framework

### PR DESCRIPTION
**Pull request by trylek**
These changes involve targeted fixes to issues seen due to the new
build mode and generic signature encoding fixes identified in
my offline investigation with JanV. While I locally see the tests
passing with just a handful of failures, I still seem unable to
make the tests run in the lab; I keep investigating that but
I believe it's useful to merge this delta in to let us parallelize
follow-up efforts.

Thanks

Tomas

### CODE REVIEW

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/InliningInfoNode.cs [hunk 1]
```diff
@@ -66,6 +66,13 @@ public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
                         continue;
                     }
 
+                    if (!factory.CompilationModuleGroup.VersionsWithMethodBody(inlinee))
+                    {
+                        // We cannot record inlining info across version bubble as cross-bubble assemblies
```
Comment (davidwrighton) [inline src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/InliningInfoNode.cs:71]: We should assert that this only happens for NonVersionable methods. (If its happening for any other method, we shouldn't actually be inclining)

**Review by davidwrighton — CHANGES_REQUESTED** [review]
One minor nit, but otherwise looks good.

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/SignatureBuilder.cs [hunk 2]
```diff
@@ -310,6 +310,12 @@ public void EmitTypeSignature(TypeDesc typeDesc, SignatureContext context)
                 case TypeFlags.ValueType:
                 case TypeFlags.Nullable:
                 case TypeFlags.Enum:
+                    if (typeDesc == PrimitiveTypeProvider.GetPrimitiveType(typeDesc.Context, PrimitiveTypeCode.TypedReference))
```
Comment (MichalStrehovsky) [inline src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/SignatureBuilder.cs:313]: Nit: this is more compact/readable (same compact form can be used for the diff in SignatureContext.cs)

```suggestion
                    if (typeDesc.IsWellKnownType(WellKnownType.TypedReference))
```

**Review by MichalStrehovsky — APPROVED** [review]

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/SignatureBuilder.cs (hunk 2 above)
Comment (trylek) [reply src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/SignatureBuilder.cs]: Thanks for pointing that out, fixed in 2nd commit along with one more place that was using the same check.

File: src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/InliningInfoNode.cs (hunk 1 above)
Comment (trylek) [reply src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/InliningInfoNode.cs:71]: Good point, fixed in 3rd commit. Could you please take another look when you have a chance?

**Review by davidwrighton — APPROVED** [review]
Looks good

**Comment by trylek** [comment]
Hit two unrelated bugs, filed issues for both:

1) Error in BinderTracingTest.Basic on all OS / architecture combos in Crossgen2 compilation mode: 

https://github.com/dotnet/runtime/issues/34509

2) Infra failure in perf test (error in CI setup script on Linux x64):

https://github.com/dotnet/runtime/issues/34511

**Pull request merged**